File: ios/fpjs_pro_plugin/Sources/fpjs_pro_plugin/FingerprintClientCache.swift
@preconcurrency import Fingerprint
import Foundation

/// One [FingerprintClientProviding] per resolved configuration.
final class FingerprintClientCache: @unchecked Sendable {
  // Fields, not a joined string. A delimiter key can collide or drop a field.
  private struct ClientKey: Hashable {
    var apiKey: String
    var regionCode: String
    var customDomain: String?
    var customFallbacks: [String]
    var pluginVersion: String
    var allowUseOfLocationData: Bool

    init(configuration: Configuration, pluginVersion: String) {
      apiKey = configuration.apiKey
      self.pluginVersion = pluginVersion
      allowUseOfLocationData = configuration.allowUseOfLocationData
      switch configuration.region {
      case .eu:
        regionCode = "eu"
        customDomain = nil
        customFallbacks = []
      case .ap:
        regionCode = "ap"
        customDomain = nil
        customFallbacks = []
      case .custom(let domain, let fallback):
        regionCode = "custom"
        customDomain = domain
        customFallbacks = fallback
      default:
        regionCode = "global"
        customDomain = nil
        customFallbacks = []
      }
    }
  }

  private var clients: [ClientKey: FingerprintClientProviding] = [:]
  private let createClient: (Configuration) -> FingerprintClientProviding
  private let lock = NSLock()

  init(
    createClient: @escaping (Configuration) -> FingerprintClientProviding = {
      FingerprintFactory.getInstance($0)
    }
  ) {
    self.createClient = createClient
  }

  func getOrCreate(
    configuration: Configuration,
    pluginVersion: String
  ) -> FingerprintClientProviding {
    lock.lock()
    defer { lock.unlock() }
    let key = ClientKey(configuration: configuration, pluginVersion: pluginVersion)
    if let existing = clients[key] {
      return existing
    }
    let client = createClient(configuration)
    clients[key] = client
    return client
  }
}

File: android/src/main/kotlin/com/fingerprint/flutter/FingerprintClientCache.kt
// Memoizes Fingerprint clients by resolved native configuration.
package com.fingerprint.flutter

import android.content.Context
import com.fingerprint.android.Configuration
import com.fingerprint.android.Fingerprint
import com.fingerprint.android.FingerprintFactory
import java.util.concurrent.ConcurrentHashMap

internal typealias FingerprintFactoryFn = (Context, Configuration) -> Fingerprint

internal class FingerprintClientCache(
  private val applicationContext: Context,
  private val createFingerprint: FingerprintFactoryFn = { context, configuration ->
    FingerprintFactory(context).createInstance(configuration)
  },
) {
  // Fields, not a joined string. A delimiter key can collide or drop a field.
  private data class ClientKey(
    val apiKey: String,
    val region: Configuration.Region,
    val endpointUrl: String,
    val fallbackEndpointUrls: List<String>,
    val pluginVersion: String,
    val allowUseOfLocationData: Boolean,
    val locationTimeoutMillis: Long,
  )

  private val clients = ConcurrentHashMap<ClientKey, Fingerprint>()

  fun getOrCreate(configuration: Configuration, pluginVersion: String): Fingerprint {
    val key = ClientKey(
      configuration.apiKey,
      configuration.region,
      configuration.endpointUrl,
      configuration.fallbackEndpointUrls.toList(),
      pluginVersion,
      configuration.allowUseOfLocationData,
      configuration.locationTimeoutMillis,
    )
    // Kotlin ConcurrentHashMap.getOrPut is not atomic: concurrent first hits can
    // each create a Fingerprint client. computeIfAbsent runs the factory once.
    // https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-put.html
    return clients.computeIfAbsent(key) {
      createFingerprint(applicationContext, configuration)
    }
  }
}
